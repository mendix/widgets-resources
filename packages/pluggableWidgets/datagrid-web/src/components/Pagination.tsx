import { createElement, Dispatch, ReactElement, SetStateAction } from "react";

export interface PaginationProps {
    canNextPage: boolean;
    canPreviousPage: boolean;
    gotoPage: (page: number) => void;
    nextPage: () => void;
    numberOfItems?: number;
    page: number;
    pageSize: number;
    previousPage: () => void;
    setPaginationIndex?: Dispatch<SetStateAction<number>>;
}

export function Pagination(props: PaginationProps): ReactElement | null {
    const numberOfPages =
        props.numberOfItems !== undefined ? Math.ceil(props.numberOfItems / props.pageSize) : undefined;
    const lastPage = numberOfPages !== undefined ? numberOfPages - 1 : 0;
    const hasLastPage = numberOfPages !== undefined;
    const initialItem = props.page * props.pageSize + 1;
    const lastItem =
        props.canNextPage || !props.numberOfItems ? (props.page + 1) * props.pageSize : props.numberOfItems;
    const setPageIndex = (page: number): void => {
        if (props.setPaginationIndex) {
            props.setPaginationIndex(page);
        }
    };

    if (props.numberOfItems === 0) {
        return null;
    }

    return (
        <div className="pagination-bar">
            <button
                className="btn pagination-button"
                disabled={props.page === 0}
                onClick={() => {
                    props.gotoPage(0);
                    setPageIndex(0);
                }}
            >
                <span className="glyphicon glyphicon-step-backward" />
            </button>
            <button
                className="btn pagination-button"
                disabled={!props.canPreviousPage}
                onClick={() => {
                    props.previousPage();
                    setPageIndex(props.page - 1);
                }}
            >
                <span className="glyphicon glyphicon-backward" />
            </button>
            <div className="paging-status">
                {initialItem} to {lastItem}{" "}
                {hasLastPage ? `of ${props.numberOfItems ?? (numberOfPages ?? 1) * props.pageSize}` : ""}
            </div>
            <button
                className="btn pagination-button"
                disabled={!props.canNextPage}
                onClick={() => {
                    props.nextPage();
                    setPageIndex(props.page + 1);
                }}
            >
                <span className="glyphicon glyphicon-forward" />
            </button>
            {hasLastPage && (
                <button
                    className="btn pagination-button"
                    disabled={props.page === lastPage}
                    onClick={() => {
                        props.gotoPage(lastPage);
                        setPageIndex(lastPage);
                    }}
                >
                    <span className="glyphicon glyphicon-step-forward" />
                </button>
            )}
        </div>
    );
}
