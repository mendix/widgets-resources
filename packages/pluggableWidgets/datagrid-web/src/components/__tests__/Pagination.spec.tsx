import { shallow } from "enzyme";
import { createElement } from "react";
import { Pagination, PaginationProps } from "../Pagination";

describe("Pagination", () => {
    it("renders the structure correctly", () => {
        const component = shallow(<Pagination {...mockPaginationProps()} />);

        expect(component).toMatchSnapshot();
    });

    it("disables previous button if is first page", () => {
        const component = shallow(<Pagination {...mockPaginationProps()} canPreviousPage={false} />);

        expect(
            component
                .find(".btn")
                .at(1)
                .prop("disabled")
        ).toBeTruthy();

        expect(component).toMatchSnapshot();
    });

    it("disables first page button if is first page", () => {
        const component = shallow(<Pagination {...mockPaginationProps()} page={0} />);

        expect(
            component
                .find(".btn")
                .first()
                .prop("disabled")
        ).toBeTruthy();

        expect(component).toMatchSnapshot();
    });

    it("disables next button if is last page", () => {
        const component = shallow(<Pagination {...mockPaginationProps()} canNextPage={false} />);

        expect(
            component
                .find(".btn")
                .at(2)
                .prop("disabled")
        ).toBeTruthy();

        expect(component).toMatchSnapshot();
    });

    it("renders the current page correctly", () => {
        const component = shallow(<Pagination {...mockPaginationProps()} />);
        expect(component.find(".paging-status").text()).toBe("1 to 5 of 25");
        component.setProps({ page: 4 });
        expect(component.find(".paging-status").text()).toBe("21 to 25 of 25");

        expect(component).toMatchSnapshot();
    });

    it("renders the current page correctly with server side paging", () => {
        const component = shallow(<Pagination {...mockPaginationProps()} numberOfItems={undefined} />);

        expect(component).toMatchSnapshot();
    });

    it("renders the current page correctly with without pages", () => {
        const component = shallow(<Pagination {...mockPaginationProps()} numberOfItems={0} />);

        expect(component).toMatchSnapshot();
    });
});

function mockPaginationProps(): PaginationProps {
    return {
        canPreviousPage: true,
        canNextPage: true,
        gotoPage: jest.fn(),
        numberOfItems: 25,
        nextPage: jest.fn(),
        page: 0,
        pageSize: 5,
        previousPage: jest.fn()
    };
}
