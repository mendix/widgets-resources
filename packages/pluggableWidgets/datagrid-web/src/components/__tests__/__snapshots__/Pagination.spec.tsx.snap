// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination disables first page button if is first page 1`] = `
<div
  className="pagination"
>
  <button
    className="btn"
    disabled={true}
    onClick={[Function]}
  >
    <span
      className="glyphicon glyphicon-step-backward"
    />
  </button>
  <button
    className="btn"
    disabled={false}
    onClick={[Function]}
  >
    <span
      className="glyphicon glyphicon-backward"
    />
  </button>
  <div
    className="paging-status"
  >
    1
     to 
    5
     
    of 25
  </div>
  <button
    className="btn"
    disabled={false}
    onClick={[Function]}
  >
    <span
      className="glyphicon glyphicon-forward"
    />
  </button>
  <button
    className="btn"
    disabled={false}
    onClick={[Function]}
  >
    <span
      className="glyphicon glyphicon-step-forward"
    />
  </button>
</div>
`;

exports[`Pagination disables next button if is last page 1`] = `
<div
  className="pagination"
>
  <button
    className="btn"
    disabled={true}
    onClick={[Function]}
  >
    <span
      className="glyphicon glyphicon-step-backward"
    />
  </button>
  <button
    className="btn"
    disabled={false}
    onClick={[Function]}
  >
    <span
      className="glyphicon glyphicon-backward"
    />
  </button>
  <div
    className="paging-status"
  >
    1
     to 
    25
     
    of 25
  </div>
  <button
    className="btn"
    disabled={true}
    onClick={[Function]}
  >
    <span
      className="glyphicon glyphicon-forward"
    />
  </button>
  <button
    className="btn"
    disabled={false}
    onClick={[Function]}
  >
    <span
      className="glyphicon glyphicon-step-forward"
    />
  </button>
</div>
`;

exports[`Pagination disables previous button if is first page 1`] = `
<div
  className="pagination"
>
  <button
    className="btn"
    disabled={true}
    onClick={[Function]}
  >
    <span
      className="glyphicon glyphicon-step-backward"
    />
  </button>
  <button
    className="btn"
    disabled={true}
    onClick={[Function]}
  >
    <span
      className="glyphicon glyphicon-backward"
    />
  </button>
  <div
    className="paging-status"
  >
    1
     to 
    5
     
    of 25
  </div>
  <button
    className="btn"
    disabled={false}
    onClick={[Function]}
  >
    <span
      className="glyphicon glyphicon-forward"
    />
  </button>
  <button
    className="btn"
    disabled={false}
    onClick={[Function]}
  >
    <span
      className="glyphicon glyphicon-step-forward"
    />
  </button>
</div>
`;

exports[`Pagination renders the current page correctly 1`] = `
<div
  className="pagination"
>
  <button
    className="btn"
    disabled={false}
    onClick={[Function]}
  >
    <span
      className="glyphicon glyphicon-step-backward"
    />
  </button>
  <button
    className="btn"
    disabled={false}
    onClick={[Function]}
  >
    <span
      className="glyphicon glyphicon-backward"
    />
  </button>
  <div
    className="paging-status"
  >
    21
     to 
    25
     
    of 25
  </div>
  <button
    className="btn"
    disabled={false}
    onClick={[Function]}
  >
    <span
      className="glyphicon glyphicon-forward"
    />
  </button>
  <button
    className="btn"
    disabled={true}
    onClick={[Function]}
  >
    <span
      className="glyphicon glyphicon-step-forward"
    />
  </button>
</div>
`;

exports[`Pagination renders the current page correctly with server side paging 1`] = `
<div
  className="pagination"
>
  <button
    className="btn"
    disabled={true}
    onClick={[Function]}
  >
    <span
      className="glyphicon glyphicon-step-backward"
    />
  </button>
  <button
    className="btn"
    disabled={false}
    onClick={[Function]}
  >
    <span
      className="glyphicon glyphicon-backward"
    />
  </button>
  <div
    className="paging-status"
  >
    1
     to 
    5
     
  </div>
  <button
    className="btn"
    disabled={false}
    onClick={[Function]}
  >
    <span
      className="glyphicon glyphicon-forward"
    />
  </button>
</div>
`;

exports[`Pagination renders the current page correctly with without pages 1`] = `
<div
  className="pagination"
>
  <button
    className="btn"
    disabled={true}
    onClick={[Function]}
  >
    <span
      className="glyphicon glyphicon-step-backward"
    />
  </button>
  <button
    className="btn"
    disabled={false}
    onClick={[Function]}
  >
    <span
      className="glyphicon glyphicon-backward"
    />
  </button>
  <div
    className="paging-status"
  >
    1
     to 
    5
     
    of 0
  </div>
  <button
    className="btn"
    disabled={false}
    onClick={[Function]}
  >
    <span
      className="glyphicon glyphicon-forward"
    />
  </button>
  <button
    className="btn"
    disabled={false}
    onClick={[Function]}
  >
    <span
      className="glyphicon glyphicon-step-forward"
    />
  </button>
</div>
`;

exports[`Pagination renders the structure correctly 1`] = `
<div
  className="pagination"
>
  <button
    className="btn"
    disabled={true}
    onClick={[Function]}
  >
    <span
      className="glyphicon glyphicon-step-backward"
    />
  </button>
  <button
    className="btn"
    disabled={false}
    onClick={[Function]}
  >
    <span
      className="glyphicon glyphicon-backward"
    />
  </button>
  <div
    className="paging-status"
  >
    1
     to 
    5
     
    of 25
  </div>
  <button
    className="btn"
    disabled={false}
    onClick={[Function]}
  >
    <span
      className="glyphicon glyphicon-forward"
    />
  </button>
  <button
    className="btn"
    disabled={false}
    onClick={[Function]}
  >
    <span
      className="glyphicon glyphicon-step-forward"
    />
  </button>
</div>
`;
